File: task/src/main/java/com/goexp/galgame/data/parser/GetchuTagParser.java
package com.goexp.galgame.data.parser;

import com.goexp.galgame.common.model.TagType;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;

import java.util.List;
import java.util.stream.Collectors;

public class GetchuTagParser {

    public List<TagType> parse(String html) {
        return Jsoup.parse(html)
                .select("#wrapper div.pc_headword:contains(カテゴリ一覧)").first()
                .parent()
                .select("div.category_pc_t")
                .stream()
                .map(this::parse)
                .collect(Collectors.toUnmodifiableList());
    }

    private TagType parse(Element item) {


        var tagType = new TagType();
        tagType.type = item.text();
        tagType.order = 0;


        tagType.tags = item.nextElementSibling().select("a")
                .stream()
                .map(ele -> ele.text().trim())
                .collect(Collectors.toUnmodifiableList());

        return tagType;

    }
}

File: task/src/main/scala/com/goexp/galgame/data/parser/GetchuTagParser.scala
package com.goexp.galgame.data.parser

import com.goexp.galgame.common.model.TagType
import org.jsoup.Jsoup
import org.jsoup.nodes.Element

import scala.jdk.CollectionConverters._


class GetchuTagParser {
  def parse(html: String): LazyList[TagType] =
    Jsoup.parse(html)
      .select("#wrapper div.pc_headword:contains(カテゴリ一覧)")
      .first
      .parent
      .select("div.category_pc_t")
      .asScala
      .to(LazyList)
      .map(parse)

  private def parse(item: Element) = {
    val list = item
      .nextElementSibling
      .select("a")
      .asScala
      .to(LazyList)
      .map {
        _.text.trim
      }
      .toList

    TagType(item.text, 0, list)

  }
}